File: src/main/java/core/packetproxy/common/FilterIO.java
package packetproxy.common;

import static packetproxy.util.Logging.errWithStackTrace;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.annotations.SerializedName;
import java.util.Collections;
import java.util.List;
import packetproxy.AppInitializer;
import packetproxy.model.*;
import packetproxy.model.Filter;

public class FilterIO {

	private static class DaoHub {

		@SerializedName(value = "filters")
		List<Filter> filterList;
	}

	public FilterIO() {
	}

	public String getOptions() throws Exception {
		DaoHub daoHub = new DaoHub();

		daoHub.filterList = AppInitializer.getFilters().queryAll();
		Collections.reverse(daoHub.filterList);

		Gson gson = new GsonBuilder().disableHtmlEscaping().setPrettyPrinting().create();
		String json = gson.toJson(daoHub);

		return json;
	}

	public void setOptions(String json) {
		try {

			DaoHub daoHub = new Gson().fromJson(json, DaoHub.class);

			Database.getInstance().dropFilters();

			for (Filter filter : daoHub.filterList) {

				Filter f = new Filter(filter.getName(), filter.getFilter());
				AppInitializer.getFilters().create(f);
			}
		} catch (Exception e) {

			errWithStackTrace(e);
		}
	}
}

File: src/main/kotlin/core/packetproxy/common/FilterIO.kt
package packetproxy.common

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.annotations.SerializedName
import packetproxy.model.Database
import packetproxy.model.Filter
import packetproxy.model.Filters
import packetproxy.util.Logging.errWithStackTrace

class FilterIO {
  private class DaoHub {
    @SerializedName("filters") lateinit var filterList: List<Filter>
  }

  @Throws(Exception::class)
  fun getOptions(): String {
    val daoHub = DaoHub()
    daoHub.filterList = Filters.getInstance().queryAll().toMutableList().also { it.reverse() }
    val gson = GsonBuilder().disableHtmlEscaping().setPrettyPrinting().create()
    return gson.toJson(daoHub)
  }

  fun setOptions(json: String) {
    try {
      val daoHub = Gson().fromJson(json, DaoHub::class.java)
      Database.getInstance().dropFilters()
      for (filter in daoHub.filterList) {
        val f = Filter(filter.getName()!!, filter.getFilter()!!)
        Filters.getInstance().create(f)
      }
    } catch (e: Exception) {
      errWithStackTrace(e)
    }
  }
}
